core/processor/abci.go

```diff
@@ -807,6 +803,55 @@ func (app *App) OnBeginBlock(
 	return ctx, resp
 }
 
+func (app *App) startProtocolUpgrade(ctx context.Context) {
+	// Stop blockchain server so it doesn't acceptc transactions and it doesn't times out.
+	go func() { app.stopBlockchain() }()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var eventsCh <-chan events.Event
+	var errsCh <-chan error
+	if app.broker.StreamingEnabled() {
+		// wait here for data node send back the confirmation
+		eventsCh, errsCh = app.broker.SocketClient().Receive(ctx)
+	}
+
+	app.broker.Send(
+		events.NewProtocolUpgradeStarted(ctx, eventspb.ProtocolUpgradeStarted{
+			LastBlockHeight: app.stats.Height(),
+		}),
+	)
+
+	if eventsCh != nil {
+		app.log.Info("waiting for data node to get ready for upgrade")
+
+	Loop:
+		for {
+			select {
+			case e := <-eventsCh:
+				if e.Type() != events.ProtocolUpgradeDataNodeReadyEvent {
+					continue
+				}
+				if e.StreamMessage().GetProtocolUpgradeDataNodeReady().GetLastBlockHeight() == app.stats.Height() {
+					cancel()
+					break Loop
+				}
+			case err := <-errsCh:
+				app.log.Fatal("failed to wait for data node to get ready for upgrade", logging.Error(err))
+			}
+		}
+	}
```

Comment:
should we have a loop here? ultimately the datanode would send only 1 event ever? so just a select might be fine?

Reply:
Well I think this is just for security. Just in case data node sends more events in the future? Double send etc?